k8s/cmd/nsm-coredns/plugin/fanout/fanout.go

```diff
@@ -67,40 +67,79 @@ func (f *Fanout) ServeDNS(ctx context.Context, w dns.ResponseWriter, m *dns.Msg)
 	timeoutContext, cancel := context.WithTimeout(ctx, defaultTimeout)
 	defer cancel()
 	clientCount := len(f.clients)
-	result := make(chan connectResult, clientCount)
+	resultCh := make(chan connectResult, clientCount)
 	for i := 0; i < clientCount; i++ {
 		client := f.clients[i]
-		go connect(timeoutContext, client, req, result, f.maxFailCount)
+		go connect(timeoutContext, client, req, resultCh, f.maxFailCount)
 	}
-	var first *connectResult
-	for first == nil {
-		select {
-		case <-timeoutContext.Done():
-			return dns.RcodeServerFailure, errContextDone
-		case conn := <-result:
-			if conn.err != nil {
-				clientCount--
-				if clientCount == 0 {
-					return dns.RcodeServerFailure, errNoHealthy
-				}
-				break
-			}
-			first = &conn
-		}
+	result := getFanoutResult(timeoutContext, clientCount, resultCh)
+	if result == nil {
+		return dns.RcodeServerFailure, errContextDone
 	}
-
-	taperr := toDnstap(ctx, first.client.addr, req, first.response, first.start)
-	if !req.Match(first.response) {
-		debug.Hexdumpf(first.response, "Wrong reply for id: %d, %s %d", first.response.Id, req.QName(), req.QType())
+	if result.err != nil {
+		return dns.RcodeServerFailure, errNoHealthy
+	}
+	taperr := toDnstap(ctx, result.client.addr, req, result.response, result.start)
+	if !req.Match(result.response) {
+		debug.Hexdumpf(result.response, "Wrong reply for id: %d, %s %d", result.response.Id, req.QName(), req.QType())
 		formerr := new(dns.Msg)
 		formerr.SetRcode(req.Req, dns.RcodeFormatError)
 		checkErr(w.WriteMsg(formerr))
 		return 0, taperr
 	}
-	checkErr(w.WriteMsg(first.response))
+	checkErr(w.WriteMsg(result.response))
 	return 0, taperr
 }
 
+func getFanoutResult(ctx context.Context, clientCount int, ch <-chan connectResult) *connectResult {
+	count := clientCount
+	var result *connectResult
+	for {
+		select {
+		case <-ctx.Done():
+			return result
+		case r := <-ch:
+			count--
+			if isBetter(result, &r) {
+				result = &r
+			}
+			if count == 0 {
+				return result
+			}
+			if r.err != nil {
+				break
+			}
+			if r.response.Rcode != dns.RcodeSuccess {
+				break
+			}
```

Comment:
Why do we need this checks? I think we can add `else` branch to 
```go
if isBetter(result, &r) {
 	result = &r
} else {
	break
}
```

Reply:
>Why do we need this checks? 

Because `isbetter` compares two connection results. After that, we need to verify that the potential result is positive to return it instantly.

Reply:
I mean the case when `isBetter` returns false. In that case we won't update `result`, so checking of `r` is useless

Reply:
if `isBetter` is false we will break on line 109-114. The problem of your proposal is if `isBetter` is true and it is not positive answer. For this case, we need to check on 109-114 liens.

Reply:
If `isBetter` is true and it is "not positive answer" then why do we put it into `result` before checks?

Reply:
because we should return a negative answer, for example,  if before the result was a connection error.

Reply:
it is important for dns

Reply:
Then I don't understand what "not positive answer" you are talking about. Seems like if `isBetter` returns `true` then we can update `result` and  go to the next `for` iteration

Reply:
positive answer is DNS message witch has RCODE=0